pydantic/_internal/_generate_schema.py

```diff
@@ -1375,13 +1376,13 @@ def _union_schema(self, union_type: Any) -> core_schema.CoreSchema:
 
     def _type_alias_type_schema(
         self,
-        obj: TypeAliasType,
+        obj: TypeAliasType | GenericAlias,
     ) -> CoreSchema:
         with self.defs.get_schema_or_ref(obj) as (ref, maybe_schema):
             if maybe_schema is not None:
                 return maybe_schema
 
-            origin = get_origin(obj) or obj
+            origin: TypeAliasType = get_origin(obj) or cast(TypeAliasType, obj)
```

Comment:
Probably best to just do:

```suggestion
            origin: TypeAliasType = get_origin(obj) or obj
```